### ISSUE TITLE
Use Polynote with Conda environment

### ISSUE THREAD

**Issue by dzanaga** [polynote/polynote#660]
Hi guys,

First of all, thanks for the terrific work, Polynote looks awesome.
For my projects I like to build my environments from a conda configuration file, so usually I have my enviroments and manage them with Conda, and use Jupyter for rapid prototyping. Polynote would be an amazing replacement because of it's features, so the ideal behavior would be to be able to start polynote from the environment, (or even having it running and just be able to select the environment, like in Jupyter you can install a kernel for each environment and select it in the interface).

Given the preface, in practice, I tried to install polynote and run it from a conda environment. 
I had to install jep in the environment and then it launched.

First problem was that I had to add the miniconda3/envs/myenv/lib folder to LD_LIBRARY_PATH in order to import packages such matplotlib/opencv/gdal/geopandas that were otherwise throwing an error, e.g. 
```
conda activate myenv
LD_LIBRARY_PATH=/path/to/miniconda3/envs/myenv/lib:$LD_LIBRARY_PATH
python polynote.py
```
This allowed me to import all libraries installed in the conda environment, and everything seemed to work well after briefly testing some things.

I then tried to plot a graph with matplotlib but no figure was rendered and no error was produced. Only empty output.

Screenshots:
![polynote](https://user-images.githubusercontent.com/8113766/68926512-ffb02480-0785-11ea-92d9-35c7006f80bd.PNG)
![polynote2](https://user-images.githubusercontent.com/8113766/68926520-03dc4200-0786-11ea-9c01-e709bed39cd1.PNG)

Shell:
```
$ python polynote.py
java -cp polynote.jar:polynote.jar:/home/user/repos/polynote/deps/polynote-runtime.jar:/home/user/repos/polynote/deps/scala-xml_2.11-1.2.0.jar:/home/user/repos/polynote/deps/scala-library-2.11.12.jar:/home/user/repos/polynote/deps/scala-collection-compat_2.11-2.1.1.jar:/home/user/repos/polynote/deps/scala-compiler-2.11.12.jar:/home/user/repos/polynote/deps/polynote-spark-runtime.jar:/home/user/repos/polynote/deps/scala-reflect-2.11.12.jar:/home/user/repos/polynote/deps/portable-scala-reflect_2.11-0.1.0.jar -Djava.library.path=/home/user/miniconda3/envs/eo/lib/python3.7/site-packages/jep polynote.Main
[INFO]  Loading configuration from config.yml
[INFO]  Loaded configuration: PolynoteConfig(Listen(8192,127.0.0.1),Storage(tmp,notebooks,Map()),List(),List(),Map(),Map(),Behavior(true,Always,List()),Security(None),UI(/))
[WARN]  Polynote allows arbitrary remote code execution, which is necessary for a notebook tool to function.
        While we'll try to improve safety by adding security measures, it will never be completely safe to
        run Polynote on your personal computer. For example:

        - It's possible that other websites you visit could use Polynote as an attack vector. Browsing the web
          while running Polynote is unsafe.
        - It's possible that remote attackers could use Polynote as an attack vector. Running Polynote on a
          computer that's accessible from the internet is unsafe.
        - Even running Polynote inside a container doesn't guarantee safety, as there will always be
          privilege escalation and container escape vulnerabilities which an attacker could leverage.

        Please be diligent about checking for new releases, as they could contain fixes for critical security
        flaws.

        Please be mindful of the security issues that Polynote causes; consult your company's security team
        before running Polynote. You are solely responsible for any breach, loss, or damage caused by running
        this software insecurely.
[zio-default-async-8-1121647253] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Service bound to address /127.0.0.1:8192
[zio-default-async-8-1121647253] INFO org.http4s.server.blaze.BlazeServerBuilder -


  _____      _                   _
 |  __ \    | |                 | |
 | |__) |__ | |_   _ _ __   ___ | |_ ___
 |  ___/ _ \| | | | | '_ \ / _ \| __/ _ \
 | |  | (_) | | |_| | | | | (_) | ||  __/
 |_|   \___/|_|\__, |_| |_|\___/ \__\___|
                __/ |
               |___/

Server running at http://127.0.0.1:8192
[zio-default-async-8-1121647253] INFO org.http4s.server.blaze.BlazeServerBuilder - http4s v0.20.6 on blaze v0.14.6 started at http://127.0.0.1:8192/
[blaze-selector-2] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39928
[blaze-selector-1] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39926
[blaze-selector-3] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39936
[blaze-selector-4] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39941
[blaze-selector-0] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39946
[blaze-selector-2] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39958
[blaze-selector-1] INFO org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:39956
[INFO]  Deploying with command:
        /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.161-0.b14.el7_4.x86_64/jre/bin/java -cp polynote.jar:polynote.jar:/home/user/repos/polynote/deps/polynote-runtime.jar:/home/user/repos/polynote/deps/scala-xml_2.11-1.2.0.jar:/home/user/repos/polynote/deps/scala-library-2.11.12.jar:/home/user/repos/polynote/deps/scala-collection-compat_2.11-2.1.1.jar:/home/user/repos/polynote/deps/scala-compiler-2.11.12.jar:/home/user/repos/polynote/deps/polynote-spark-runtime.jar:/home/user/repos/polynote/deps/scala-reflect-2.11.12.jar:/home/user/repos/polynote/deps/portable-scala-reflect_2.11-0.1.0.jar -Djava.library.path=/home/user/miniconda3/envs/eo/lib/python3.7/site-packages/jep polynote.kernel.remote.RemoteKernelClient --address 127.0.0.1 --port 41185 --kernelFactory polynote.kernel.LocalKernelFactory
[INFO]  Connected. Decoding incoming messages
```

Any idea of what could be the issue?

**Comment by jonathanindig** [comment]
Hi @dzanaga thanks for trying out Polynote!

1. Just to make sure I understand: are you setting those dependencies (matplotlib, numpy, etc) in Polynote itself (e.g., in the Configuration & Dependencies section of the notebook)? 

2. What do you see when you run the following: 
```python
import matplotlib
matplotlib.get_backend()
```

**Comment by dzanaga** [comment]
Hi @jonathanindig thanks for getting back to me.

To answer your questions:
1. No, I am not setting those dependencies in polynote. I have everything installed in the environment that I use to start polynote.

2. When I run what you suggest I get:
```
Out: String
module://polynote
```
And if I run:
```
matplotlib.rcsetup.all_backends
```
I get nothing out.

Again if I do:
```
matplotlib.use('WRONG_ARG')
```
Nothing happens, while I would expect an error.

I guess I am probably doing something wrong since I am not configuring polynote dependencies, but for me it would be just great to be able to use the tool with my own conda environment and manage the environment and dependecies myself rather than having polynote taking care of it. Is that a possibility?

**Comment by jonathanindig** [comment]
You should definitely be able to use external dependencies with Polynote. It looks like your matplotlib backend has correctly been configured to use ours which should mean your figures get displayed. I agree that `matplotlib.use('WRONG_ARG')` should fail and it's definitely suspicious that it doesn't. 

Can you share a bit more information about your setup? OS, python version, etc? I'm a bit surprised that you needed to set LD_LIBRARY_PATH yourself (since the startup script is supposed to handle that) - can you share the errors you were running into before you did that? Finally, can you share your `sys.path` inside polynote?

**Comment by jonathanindig** [comment]
Oh, also, can you tell me the output of 
```python
print(plt._show)
```

I'm expecting something like the following, let me know whether you see the same. 
```
<bound method PolynoteBackend.show of <class 'polynote.PolynoteBackend'>>
```

**Comment by dzanaga** [comment]
Indeed, that's exactly what I get.
```
<bound method PolynoteBackend.show of <class 'polynote.PolynoteBackend'>>
```

**Comment by dzanaga** [comment]
Great to know! Ok, more info on my system.
I work on a VM of my company cluster. It's a CentOS 7 system. It comes with a default configuration from the IT guys, with a python 2 and python 3 installation and some packages, among which GDAL (Geospatial tools library).

Out of the box:
```
$ which python
/opt/rh/python27/root/usr/bin/python

$ echo $LD_LIBRARY_PATH
/opt/rh/python27/root/usr/lib64
```
But then I installed conda in order to manage different environments to `/home/dzanaga/miniconda3`

Inside I have an environment (eo) with GDAL and other libraries needed for my project.
I then run:
```
conda activate eo
python polynote.py
```
In the notebook if I do:
```
import matplotlib

Uncaught exception: ImportError: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/kiwisolver.cpython-37m-x86_64-linux-gnu.so) (java.lang.RuntimeException)
<string>.__polynote_run__(<string>:54)
<ast>.<module>(<ast>:1)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/matplotlib/__init__.py:200)
__init__.py._check_versions(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/matplotlib/__init__.py:194)
__init__.py.import_module(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/importlib/__init__.py:127)
```

```
import numpy
(this works)
```

```
import rasterio
Uncaught exception: ImportError: /lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by /home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/rasterio/../../../libgdal.so.20) (java.lang.RuntimeException)
<string>.__polynote_run__(<string>:54)
<ast>.<module>(<ast>:1)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/rasterio/__init__.py:22)
```

```
import geopandas
Uncaught exception: ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/_libs/window.cpython-37m-x86_64-linux-gnu.so) (java.lang.RuntimeException)
<string>.__polynote_run__(<string>:54)
<ast>.<module>(<ast>:1)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/geopandas/__init__.py:1)
geoseries.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/geopandas/geoseries.py:6)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/__init__.py:55)
api.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/api.py:24)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/groupby/__init__.py:1)
generic.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/groupby/generic.py:44)
frame.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/frame.py:115)
series.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/series.py:4889)
generic.py._add_series_or_dataframe_operations(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/generic.py:10747)
window.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/window.py:13)
```

```
import pandas
Uncaught exception: AttributeError: module 'pandas' has no attribute 'plotting' (java.lang.RuntimeException)
<string>.__polynote_run__(<string>:54)
<ast>.<module>(<ast>:1)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/__init__.py:55)
api.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/api.py:24)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/groupby/__init__.py:1)
generic.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/groupby/generic.py:44)
frame.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/frame.py:115)
series.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/series.py:141)
series.py.Series(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/pandas/core/series.py:4872)
```

```
Uncaught exception: ImportError: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/scipy/sparse/_sparsetools.cpython-37m-x86_64-linux-gnu.so) (java.lang.RuntimeException)
<string>.__polynote_run__(<string>:54)
<ast>.<module>(<ast>:1)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/__init__.py:127)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/util/__init__.py:12)
_montage.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/util/_montage.py:2)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/exposure/__init__.py:1)
exposure.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/exposure/exposure.py:3)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/color/__init__.py:1)
colorconv.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/skimage/color/colorconv.py:55)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/scipy/linalg/__init__.py:213)
_sketches.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/scipy/linalg/_sketches.py:11)
__init__.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/scipy/sparse/__init__.py:230)
csr.py.<module>(/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages/scipy/sparse/csr.py:13)
```

If before starting polynote I run 
```
LD_LIBRARY_PATH=/home/dzanaga/miniconda3/envs/eo3/lib:$LD_LIBRARY_PATH
```
Then I don't get errors from the imports.

Regarding the output of `plt._show` I get exactly what you wrote.

Finally, my sys.path inside the polynote notebook is:
```
['/usr/hdp/current/spark-client/python', '/usr/hdp/current/spark-client/python/lib/py4j-0.9-src.zip', '/home/dzanaga/miniconda3/envs/eo/lib/python37.zip', '/home/dzanaga/miniconda3/envs/eo/lib/python3.7', '/home/dzanaga/miniconda3/envs/eo/lib/python3.7/lib-dynload', '/home/dzanaga/.local/lib/python3.7/site-packages', '/home/dzanaga/miniconda3/envs/eo/lib/python3.7/site-packages', '/home/dzanaga/repos/ipython']
```

and `sys.version_info`:
```
sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)
```

**Comment by jonathanindig** [comment]
Hmm, I'm a little stumped here, there's something different about your environment but I'm not sure what it is. 

What value do you get from `matplotlib.__version__`? Also, does anything different happen if you call `plt._show()` rather than `plt.show()` in your original plotting cell?

**Comment by dzanaga** [comment]
Matplotlib version is 3.1.0.

Using `plt._show()` indeed displays the plot.
I did some further testing, created a new environment with only matplotlib and the plotting was working normally.
I then went back to the previous environment, and create a new notebook, and again the plotting was working normally now.
So basically, `plt.show()` doesn't display the plot only on the first notebook that I created.

See below:
<img width="960" alt="polynote_plot1" src="https://user-images.githubusercontent.com/8113766/69322081-819bc400-0c44-11ea-9611-9c83c8fe5438.PNG">

and this in the second test notebook, working as expected:
<img width="960" alt="polynote_plot2" src="https://user-images.githubusercontent.com/8113766/69322110-90827680-0c44-11ea-8e4c-b7c3600429f0.PNG">

Thanks for your help. Closing this as indeed the bug was only related to that first notebook I created and cannot even reproduce now that I deleted it.
